calyx-opt/src/passes/static_promotion.rs

```diff
@@ -523,31 +642,86 @@ impl Visitor for StaticPromotion {
     fn finish_seq(
         &mut self,
         s: &mut ir::Seq,
-        _comp: &mut ir::Component,
-        _sigs: &LibrarySignatures,
+        comp: &mut ir::Component,
+        sigs: &LibrarySignatures,
         _comps: &[ir::Component],
     ) -> VisResult {
-        if let Some(sseq) = s.make_static() {
-            return Ok(Action::change(ir::Control::Static(
-                ir::StaticControl::Seq(sseq),
-            )));
+        let mut builder = ir::Builder::new(comp, sigs);
+        let mut new_stmts: Vec<ir::Control> = Vec::new();
+        let mut static_vec: Vec<ir::Control> = Vec::new();
+        let mut promote_control = true;
+        for stmt in std::mem::take(&mut s.stmts) {
+            if stmt.is_static()
+                || stmt.has_attribute(ir::NumAttr::PromoteStatic)
+            {
+                static_vec.push(stmt);
+            } else {
+                promote_control = false;
+                match static_vec.len().cmp(&1) {
+                    std::cmp::Ordering::Equal => {
+                        new_stmts.extend(static_vec);
+                    }
+                    std::cmp::Ordering::Greater => {
+                        let sseq = self.construct_static_seq(
+                            &mut builder,
+                            &mut static_vec,
+                        );
+                        new_stmts.push(sseq);
+                    }
+                    _ => {}
+                }
+                new_stmts.push(stmt);
+                static_vec = Vec::new();
+            }
         }
-        Ok(Action::Continue)
+        if !static_vec.is_empty() {
+            if static_vec.len() == 1 {
+                new_stmts.extend(static_vec);
+            } else {
+                let sseq =
+                    self.construct_static_seq(&mut builder, &mut static_vec);
+                if promote_control {
+                    return Ok(Action::change(sseq));
+                } else {
+                    new_stmts.push(sseq);
+                }
+            }
+        }
+        let new_seq = ir::Control::Seq(ir::Seq {
+            stmts: new_stmts,
+            attributes: ir::Attributes::default(),
+        });
+        Ok(Action::change(new_seq))
     }
 
     fn finish_par(
         &mut self,
         s: &mut ir::Par,
-        _comp: &mut ir::Component,
-        _sigs: &LibrarySignatures,
+        comp: &mut ir::Component,
+        sigs: &LibrarySignatures,
         _comps: &[ir::Component],
     ) -> VisResult {
-        if let Some(spar) = s.make_static() {
-            return Ok(Action::change(ir::Control::Static(
-                ir::StaticControl::Par(spar),
-            )));
+        let mut builder = ir::Builder::new(comp, sigs);
+        let mut new_stmts: Vec<ir::Control> = Vec::new();
+        let (mut s_stmts, d_stmts): (Vec<ir::Control>, Vec<ir::Control>) =
+            std::mem::take(&mut s.stmts).into_iter().partition(|s| {
```

Comment:
I'm not sure what's better, but fwiw I usually use `s.stmts.drain(..)` in these situations instead of `std::mem::take` and `into_iter`. Not that it matters too much and idk if `drain(..)` is even better.